File: tools/compiler.py
#!/usr/bin/env python3

"""
Compiler Tool - Python Implementation
Compiles C/C++ source code using GCC/G++ with configurable flags
"""

import json
import sys
import os
import argparse
import subprocess
import tempfile
from pathlib import Path
from typing import Dict, List, Any

class Compiler:
    def __init__(self, source_path: str, output_path: str = '',
                 compiler: str = 'gcc', flags: List[str] = None,
                 optimization: str = 'O0', arch: str = 'x86_64'):
        self.source_path = source_path
        self.output_path = output_path or Path(source_path).stem
        self.compiler = compiler
        self.flags = flags or []
        self.optimization = f'-{optimization}' if not optimization.startswith('-') else optimization
        self.arch = arch

    def compile(self) -> Dict[str, Any]:
        """Compile the source file"""
        try:
            if not Path(self.source_path).exists():
                raise FileNotFoundError(f"Source file not found: {self.source_path}")

            ext = Path(self.source_path).suffix.lower()
            if ext not in ('.c', '.cpp', '.cc', '.cxx', '.s', '.asm'):
                raise ValueError(f"Unsupported source file extension: {ext}")

            # Determine default compiler by extension if not overridden
            compiler = self.compiler
            if compiler == 'gcc' and ext in ('.cpp', '.cc', '.cxx'):
                compiler = 'g++'

            # Build command
            cmd = [compiler, self.source_path, '-o', self.output_path, self.optimization]

            # Add architecture flags
            if self.arch == 'x86':
                cmd.extend(['-m32'])
            elif self.arch == 'x86_64':
                cmd.extend(['-m64'])
            elif self.arch == 'arm':
                cmd.extend(['-marm'])
            elif self.arch == 'arm64':
                cmd.extend(['-march=armv8-a'])

            # Add extra flags
            cmd.extend(self.flags)

            # Run compilation
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=60
            )

            if result.returncode == 0:
                return {
                    'output_path': os.path.abspath(self.output_path),
                    'compiler': compiler,
                    'flags': cmd[1:],
                    'arch': self.arch,
                    'warnings': result.stderr.strip(),
                    'binary_size': os.path.getsize(self.output_path) if os.path.exists(self.output_path) else 0,
                    'summary': f'Compilation successful: {self.source_path} -> {self.output_path}'
                }
            else:
                return {
                    'error': result.stderr.strip(),
                    'compiler': compiler,
                    'flags': cmd[1:],
                    'summary': f'Compilation failed: {result.stderr.strip()}'
                }

        except Exception as e:
            return {'error': str(e), 'summary': f'Compilation error: {str(e)}'}

    def get_file_info(self) -> Dict[str, Any]:
        """Get information about the compiled binary"""
        try:
            if not os.path.exists(self.output_path):
                return {}

            result = subprocess.run(
                ['file', self.output_path],
                capture_output=True,
                text=True,
                timeout=10
            )

            return {
                'file_type': result.stdout.strip(),
                'size_bytes': os.path.getsize(self.output_path),
            }

        except Exception as e:
            return {'error': str(e)}


def main():
    parser = argparse.ArgumentParser(description='Compiler Tool')
    parser.add_argument('--source', required=True, help='Path to source file')
    parser.add_argument('--output', help='Output binary path')
    parser.add_argument('--compiler', choices=['gcc', 'g++'], default='gcc', help='Compiler to use')
    parser.add_argument('--flags', default='', help='Additional compiler flags (space-separated)')
    parser.add_argument('--optimization', choices=['O0', 'O1', 'O2', 'O3', 'Os'], default='O0',
                        help='Optimization level')
    parser.add_argument('--arch', choices=['x86', 'x86_64', 'arm', 'arm64'], default='x86_64',
                        help='Target architecture')
    parser.add_argument('--output-format', choices=['json', 'text'], default='json')

    args = parser.parse_args()

    try:
        flags = [f.strip() for f in args.flags.split() if f.strip()] if args.flags else []

        compiler = Compiler(
            source_path=args.source,
            output_path=args.output or '',
            compiler=args.compiler,
            flags=flags,
            optimization=args.optimization,
            arch=args.arch
        )

        compile_result = compiler.compile()
        file_info = compiler.get_file_info() if compile_result.get('output_path') else {}

        output = {
            'data': {
                'compilation': compile_result,
                'file_info': file_info
            },
            'summary': compile_result.get('summary', 'Compilation completed'),
            'metadata': {
                'source': args.source,
                'output': args.output or Path(args.source).stem,
                'compiler': args.compiler,
                'optimization': args.optimization,
                'arch': args.arch,
                'ai_next_step': 'Use compiled binary for BinDiff analysis or exploit testing'
            }
        }

        if args.output_format == 'json':
            print(json.dumps(output, indent=2))
        else:
            print(f"Compiler Report")
            print(f"==============")
            print(f"Source: {args.source}")
            print(f"Output: {output.get('data', {}).get('compilation', {}).get('output_path', args.output or Path(args.source).stem)}")
            print(f"Status: {'Success' if compile_result.get('output_path') else 'Failed'}")
            if compile_result.get('output_path'):
                print(f"Size: {file_info.get('size_bytes', 0)} bytes")
            if compile_result.get('error'):
                print(f"Error: {compile_result['error']}")

    except Exception as e:
        error_output = {
            'error': str(e),
            'summary': f'Compiler failed: {str(e)}',
            'metadata': {'ai_next_step': 'Check source file and GCC/G++ installation'}
        }
        print(json.dumps(error_output, indent=2))
        sys.exit(1)


if __name__ == '__main__':
    main()

File: tools/compiler.ts
#!/usr/bin/env node

/**
 * Compiler Tool - TypeScript Wrapper
 * Compiles C/C++ source code with GCC/G++ for testing and exploit verification
 */

import * as fs from 'fs';
import * as path from 'path';
import * as childProcess from 'child_process';

interface CompilerConfig {
  source_path: string;
  output_path?: string;
  compiler?: 'gcc' | 'g++';
  flags?: string[];
  optimization_level?: 'O0' | 'O1' | 'O2' | 'O3' | 'Os';
  arch?: string;
  output_format?: 'json' | 'text';
}

interface ToolResult {
  status: 'success' | 'error';
  tool_name: string;
  execution_time_ms: number;
  data: any;
  summary: string;
  metadata: {
    ai_next_step?: string;
    [key: string]: any;
  };
}

async function runCompiler(config: CompilerConfig): Promise<ToolResult> {
  const startTime = Date.now();

  try {
    if (!config.source_path) {
      throw new Error('source_path is required');
    }

    if (!fs.existsSync(config.source_path)) {
      throw new Error(`Source file not found: ${config.source_path}`);
    }

    const pythonScriptPath = path.join(__dirname, 'compiler.py');
    const args = [
      pythonScriptPath,
      '--source', config.source_path,
    ];

    if (config.output_path) args.push('--output', config.output_path);
    if (config.compiler) args.push('--compiler', config.compiler);
    if (config.flags) args.push('--flags', config.flags.join(' '));
    if (config.optimization_level) args.push('--optimization', config.optimization_level);
    if (config.arch) args.push('--arch', config.arch);
    args.push('--output-format', config.output_format || 'json');

    return new Promise((resolve) => {
      const proc = childProcess.spawn('python3', args, {
        stdio: ['pipe', 'pipe', 'pipe']
      });

      let stdout = '';
      let stderr = '';

      proc.stdout?.on('data', (data) => { stdout += data.toString(); });
      proc.stderr?.on('data', (data) => { stderr += data.toString(); });

      proc.on('close', (code) => {
        const executionTime = Date.now() - startTime;

        if (code !== 0) {
          const result: ToolResult = {
            status: 'error',
            tool_name: 'compiler',
            execution_time_ms: executionTime,
            data: { stderr },
            summary: `Compilation failed: ${stderr || stdout}`,
            metadata: { ai_next_step: 'Fix source code errors and retry' }
          };
          try {
            const parsed = JSON.parse(stdout);
            if (parsed.error || parsed.summary) {
              result.data = parsed;
              result.summary = parsed.summary || result.summary;
            }
          } catch (_) {}
          resolve(result);
          return;
        }

        try {
          const result = JSON.parse(stdout);
          result.status = 'success';
          result.tool_name = 'compiler';
          result.execution_time_ms = executionTime;
          result.metadata = result.metadata || {};
          result.metadata.ai_next_step = 'Binary compiled; pass to BinDiff or test with exploit';
          resolve(result);
        } catch (e) {
          resolve({
            status: 'error',
            tool_name: 'compiler',
            execution_time_ms: executionTime,
            data: { stdout, stderr },
            summary: 'Failed to parse compiler output',
            metadata: { ai_next_step: 'Check Python script implementation' }
          });
        }
      });
    });
  } catch (error) {
    const executionTime = Date.now() - startTime;
    const message = error instanceof Error ? error.message : String(error);
    return {
      status: 'error',
      tool_name: 'compiler',
      execution_time_ms: executionTime,
      data: { error: message },
      summary: `Compiler error: ${message}`,
      metadata: { ai_next_step: 'Fix configuration and retry' }
    };
  }
}

async function main() {
  const args = process.argv.slice(2);
  const config: CompilerConfig = {
    source_path: '',
    output_format: 'json',
    compiler: 'gcc',
    optimization_level: 'O0'
  };

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--source': config.source_path = args[++i]; break;
      case '--output': config.output_path = args[++i]; break;
      case '--compiler': config.compiler = args[++i] as 'gcc' | 'g++'; break;
      case '--flags': config.flags = args[++i].split(' '); break;
      case '--optimization': config.optimization_level = args[++i] as any; break;
      case '--arch': config.arch = args[++i]; break;
      case '--output-format': config.output_format = args[++i] as 'json' | 'text'; break;
    }
  }

  const result = await runCompiler(config);
  console.log(JSON.stringify(result, null, 2));
}

main().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});
